compiler/src/dotty/tools/dotc/transform/MacroAnnotations.scala

```diff
@@ -31,82 +31,100 @@ class MacroAnnotations(phase: IdentityDenotTransformer):
 
-      val allTrees = List.newBuilder[DefTree]
-      var insertedAfter: List[List[DefTree]] = Nil
+      val allTrees = List.newBuilder[MemberDef]
+      var insertedAfter: List[List[MemberDef]] = Nil
 
       // Apply all macro annotation to `tree` and collect new definitions in order
-      val transformedTree: DefTree = tree.symbol.annotations.foldLeft(tree) { (tree, annot) =>
-        if isMacroAnnotation(annot) then
-          debug.println(i"Expanding macro annotation: ${annot}")
-
-          // Interpret call to `new myAnnot(..).transform(using <Quotes>)(<tree>)`
-          val transformedTrees = callMacro(macroInterpreter, tree, annot)
-          transformedTrees.span(_.symbol != tree.symbol) match
-            case (prefixed, newTree :: suffixed) =>
-              allTrees ++= prefixed
-              insertedAfter = suffixed :: insertedAfter
-              for prefixedTree <- prefixed do
-                checkMacroDef(prefixedTree, tree, annot)
-              for suffixedTree <- suffixed do
-                checkMacroDef(suffixedTree, tree, annot)
-              TreeChecker.checkMacroGeneratedTree(tree, newTree)
-              newTree
-            case (Nil, Nil) =>
-              report.error(i"Unexpected `Nil` returned by `(${annot.tree}).transform(..)` during macro expansion", annot.tree.srcPos)
-              tree
-            case (_, Nil) =>
-              report.error(i"Transformed tree for ${tree} was not return by `(${annot.tree}).transform(..)` during macro expansion", annot.tree.srcPos)
-              tree
-        else
-          tree
-      }
-
+      val (transformedTree, transformedCompanion) =
+        tree.symbol.annotations.foldLeft((tree, companion)): (lastResult, annot) =>
+          val (tree, companion) = lastResult
+          if isMacroAnnotation(annot) then
+            debug.println(i"Expanding macro annotation: ${annot}")
+            // Interpret call to `new myAnnot(..).transform(using <Quotes>)(<tree>, <companion>)`
+            val (transformedTrees, transformedCompanion) = callMacro(macroInterpreter, tree, companion, annot)
+            transformedTrees.span(_.symbol != tree.symbol) match
+              case (prefixed, newTree :: suffixed) =>
+                allTrees ++= prefixed
+                insertedAfter = suffixed :: insertedAfter
+                for prefixedTree <- prefixed do
+                  checkMacroDef(prefixedTree, tree, annot)
+                for suffixedTree <- suffixed do
+                  checkMacroDef(suffixedTree, tree, annot)
+                // Check the generate tree structure
+                TreeChecker.checkMacroGeneratedTree(tree, newTree)
+                (newTree, transformedCompanion)
+              case (Nil, Nil) =>
+                report.error(i"Unexpected `Nil` returned by `(${annot.tree}).transform(..)` during macro expansion", annot.tree.srcPos)
+                (tree, transformedCompanion)
```

Comment:
In case the transformed tree is missing when expanding an annotation, we still transform the others. Should the current expansion be a no-op (therefore returning the `companion` and not the `transformedComapnion`) ?
It seems that returning the `companion` is the better choice for the specification.